src/transformers/pipelines/automatic_mask_generation.py

```diff
@@ -0,0 +1,615 @@
+    """Generates a 2D grid of points evenly spaced in [0,1]x[0,1]."""
+    offset = 1 / (2 * n_per_side)
+    points_one_side = np.linspace(offset, 1 - offset, n_per_side)
+    points_x = np.tile(points_one_side[None, :], (n_per_side, 1))
+    points_y = np.tile(points_one_side[:, None], (1, n_per_side))
+    points = np.stack([points_x, points_y], axis=-1).reshape(-1, 2)
+    return points
+
+
+def _normalize_coordinates(target_size, coords: np.ndarray, original_size, is_bounding_box=False) -> np.ndarray:
+    """
+    Expects a numpy array of length 2 in the final dimension. Requires the original image size in (H, W) format.
+    """
+    old_h, old_w = original_size
+
+    scale = target_size * 1.0 / max(old_h, old_w)
+    new_h, new_w = old_h * scale, old_w * scale
+    new_w = int(new_w + 0.5)
+    new_h = int(new_h + 0.5)
+
+    coords = deepcopy(coords).astype(float)
+
+    if is_bounding_box:
+        # reshape to .reshape(-1, 2, 2)
+        coords = coords.reshape(-1, 2, 2)
+
+    coords[..., 0] = coords[..., 0] * (new_w / old_w)
+    coords[..., 1] = coords[..., 1] * (new_h / old_h)
+
+    if is_bounding_box:
+        # reshape back to .reshape(-1, 4)
+        coords = coords.reshape(-1, 4)
+
+    return coords
+
+
+def _generate_crop_boxes(
+    image,
+    target_size,
+    n_layers: int = 0,
+    overlap_ratio: float = 512 / 1500,
+    points_per_crop: Optional[int] = 32,
+    scale_per_layer: Optional[List[int]] = 1,
+) -> Tuple[List[List[int]], List[int]]:
+    """
+    Generates a list of crop boxes of different sizes. Each layer has (2**i)**2 boxes for the ith layer.
+    """
+    if isinstance(image, list):
+        raise ValueError("Only one image is allowed for crop generation.")
+    image = to_numpy_array(image)
+    original_size = image.shape[:2]
+
+    points_grid = []
+    for i in range(n_layers + 1):
+        n_points = int(points_per_crop / (scale_per_layer**i))
+        points_grid.append(_build_point_grid(n_points))
+
+    crop_boxes, layer_idxs = [], []
+    im_h, im_w = original_size
+    short_side = min(im_h, im_w)
+
+    # Original image
+    crop_boxes.append([0, 0, im_w, im_h])
+    layer_idxs.append(0)
+
+    for i_layer in range(n_layers):
+        n_crops_per_side = 2 ** (i_layer + 1)
+        overlap = int(overlap_ratio * short_side * (2 / n_crops_per_side))
+
+        crop_w = int(math.ceil((overlap * (n_crops_per_side - 1) + im_w) / n_crops_per_side))
+        crop_h = int(math.ceil((overlap * (n_crops_per_side - 1) + im_h) / n_crops_per_side))
+
+        crop_box_x0 = [int((crop_w - overlap) * i) for i in range(n_crops_per_side)]
+        crop_box_y0 = [int((crop_h - overlap) * i) for i in range(n_crops_per_side)]
+
+        # Crops in XYWH format
+        for x0, y0 in product(crop_box_x0, crop_box_y0):
+            box = [x0, y0, min(x0 + crop_w, im_w), min(y0 + crop_h, im_h)]
+            crop_boxes.append(box)
+            layer_idxs.append(i_layer + 1)
```

Comment:
Same comment as in the same PR. Use more precise variables as we don't know what (x,y) is. Same XYWH format could have multiple interpretations (is XY the center? the top-left corner?) so make the comment more explicit.